File: composeApp/src/androidMain/kotlin/com/humblesolutions/indsphinx/repository/BackendVisitorPassRepository.kt
package com.humblesolutions.indsphinx.repository

import com.google.firebase.Timestamp
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.Query
import com.humblesolutions.indsphinx.model.VisitorPass
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.tasks.await
import java.util.Date

class BackendVisitorPassRepository : VisitorPassRepository {
    private val db = FirebaseFirestore.getInstance()

    override suspend fun submitVisitorPass(pass: VisitorPass): String {
        val data = hashMapOf(
            "OccupantId" to pass.occupantId,
            "OccupantName" to pass.occupantName,
            "flatId" to pass.flatId,
            "FlatNumber" to pass.flatNumber,
            "VisitorName" to pass.visitorName,
            "VisitorPhone" to pass.visitorPhone,
            "PurposeOfVisit" to pass.purposeOfVisit,
            "RelationshipWithVisitor" to pass.relationshipWithVisitor,
            "VisitDate" to Timestamp(Date(pass.visitDate)),
            "RequestDate" to Timestamp(Date()),
            "Status" to "PENDING"
        )
        val doc = db.collection("VisitorPass").add(data).await()
        return doc.id
    }

    override fun observeByOccupant(occupantId: String): Flow<List<VisitorPass>> = callbackFlow {
        val registration = db.collection("VisitorPass")
            .whereEqualTo("OccupantId", occupantId)
            .orderBy("RequestDate", Query.Direction.DESCENDING)
            .addSnapshotListener { snapshot, error ->
                if (error != null || snapshot == null) { trySend(emptyList()); return@addSnapshotListener }
                val passes = snapshot.documents.mapNotNull { doc ->
                    val visitorName = doc.getString("VisitorName") ?: return@mapNotNull null
                    VisitorPass(
                        id = doc.id,
                        occupantId = doc.getString("OccupantId") ?: "",
                        occupantName = doc.getString("OccupantName") ?: "",
                        flatId = doc.getString("flatId") ?: "",
                        flatNumber = doc.getString("FlatNumber") ?: "",
                        visitorName = visitorName,
                        visitorPhone = doc.getString("VisitorPhone") ?: "",
                        purposeOfVisit = doc.getString("PurposeOfVisit") ?: "",
                        relationshipWithVisitor = doc.getString("RelationshipWithVisitor") ?: "",
                        visitDate = (doc.get("VisitDate") as? Timestamp)?.toDate()?.time ?: 0L,
                        requestDate = (doc.get("RequestDate") as? Timestamp)?.toDate()?.time ?: 0L,
                        status = doc.getString("Status") ?: "PENDING"
                    )
                }
                trySend(passes)
            }
        awaitClose { registration.remove() }
    }
}

File: iosApp/iosApp/BackendVisitorPassRepository.swift
import Foundation
import FirebaseFirestore

struct VisitorPass: Identifiable, Hashable {
    let id: String
    let occupantId: String
    let occupantName: String
    let flatId: String
    let flatNumber: String
    let visitorName: String
    let visitorPhone: String
    let purposeOfVisit: String
    let relationshipWithVisitor: String
    let visitDate: Date
    let requestDate: Date
    let status: String  // PENDING | ACCEPTED | REJECTED
}

class BackendVisitorPassRepository {
    private let db = Firestore.firestore()

    func submitVisitorPass(
        occupantId: String,
        occupantName: String,
        flatId: String,
        flatNumber: String,
        visitorName: String,
        visitorPhone: String,
        purposeOfVisit: String,
        relationshipWithVisitor: String,
        visitDate: Date
    ) async throws -> String {
        let data: [String: Any] = [
            "OccupantId": occupantId,
            "OccupantName": occupantName,
            "flatId": flatId,
            "FlatNumber": flatNumber,
            "VisitorName": visitorName,
            "VisitorPhone": visitorPhone,
            "PurposeOfVisit": purposeOfVisit,
            "RelationshipWithVisitor": relationshipWithVisitor,
            "VisitDate": Timestamp(date: visitDate),
            "RequestDate": Timestamp(date: Date()),
            "Status": "PENDING"
        ]
        let ref = try await db.collection("VisitorPass").addDocument(data: data)
        return ref.documentID
    }

    func observeByOccupant(occupantId: String, onChange: @escaping ([VisitorPass]) -> Void) -> ListenerRegistration {
        return db.collection("VisitorPass")
            .whereField("OccupantId", isEqualTo: occupantId)
            .order(by: "RequestDate", descending: true)
            .addSnapshotListener { snapshot, _ in
                guard let snapshot = snapshot else { onChange([]); return }
                let passes: [VisitorPass] = snapshot.documents.compactMap { doc in
                    let data = doc.data()
                    guard let visitorName = data["VisitorName"] as? String else { return nil }
                    return VisitorPass(
                        id: doc.documentID,
                        occupantId: data["OccupantId"] as? String ?? "",
                        occupantName: data["OccupantName"] as? String ?? "",
                        flatId: data["flatId"] as? String ?? "",
                        flatNumber: data["FlatNumber"] as? String ?? "",
                        visitorName: visitorName,
                        visitorPhone: data["VisitorPhone"] as? String ?? "",
                        purposeOfVisit: data["PurposeOfVisit"] as? String ?? "",
                        relationshipWithVisitor: data["RelationshipWithVisitor"] as? String ?? "",
                        visitDate: (data["VisitDate"] as? Timestamp)?.dateValue() ?? Date(),
                        requestDate: (data["RequestDate"] as? Timestamp)?.dateValue() ?? Date(),
                        status: data["Status"] as? String ?? "PENDING"
                    )
                }
                onChange(passes)
            }
    }
}
